src/esp/assets/BaseMesh.cpp

```diff
@@ -19,5 +24,102 @@ bool BaseMesh::setMeshType(SupportedMeshType type) {
+    }
+  } else {
+    Mn::Math::packInto(Cr::Containers::arrayCast<2, float>(
+                           stridedArrayView(srcMeshData.colorsAsArray()))
+                           .except({0, 1}),
+                       Cr::Containers::arrayCast<2, Mn::UnsignedByte>(
+                           stridedArrayView(meshColors)));
+  }
+}  // BaseMesh::buildMeshColors
+
+void BaseMesh::buildSemanticOBBs(
+    const std::vector<vec3f>& vertices,
+    const std::vector<uint16_t>& vertSemanticIDs,
+    const std::vector<std::shared_ptr<esp::scene::SemanticObject>>& ssdObjs,
+    std::vector<std::string>& debugMsgs) const {
+  // build per-SSD object vector of known semantic IDs
+  std::size_t numSSDObjs = ssdObjs.size();
+  // no semantic ID 0
+  std::vector<int> semanticIDToSSOBJidx(numSSDObjs + 1);
+  for (int i = 0; i < numSSDObjs; ++i) {
+    const auto& ssdObj = *ssdObjs[i];
+    int semanticID = ssdObj.semanticID();
+    if (semanticIDToSSOBJidx.size() <= semanticID) {
+      semanticIDToSSOBJidx.resize(semanticID + 1);
+    }
+    semanticIDToSSOBJidx[semanticID] = i;
+  }
+
+  // aggegates of per-semantic ID mins and maxes
+  std::vector<esp::vec3f> vertMax(
+      semanticIDToSSOBJidx.size(),
+      {-Mn::Constants::inf(), -Mn::Constants::inf(), -Mn::Constants::inf()});
+  std::vector<esp::vec3f> vertMin(
+      semanticIDToSSOBJidx.size(),
+      {Mn::Constants::inf(), Mn::Constants::inf(), Mn::Constants::inf()});
+  std::vector<int> vertCounts(semanticIDToSSOBJidx.size());
+
+  // for each vertex, map vert min and max for each known semantic ID
+  // Known semantic IDs are expected to be contiguous, and correspond to the
+  // number of unique ssdObjs mappings.
+
+  for (int vertIdx = 0; vertIdx < vertSemanticIDs.size(); ++vertIdx) {
+    // semantic ID on vertex - valid values are 1->semanticIDToSSOBJidx.size().
+    // Invalid/unknown semantic ids are > semanticIDToSSOBJidx.size()
+    const auto semanticID = vertSemanticIDs[vertIdx];
+    if ((semanticID > 0) && (semanticID <= semanticIDToSSOBJidx.size())) {
+      const auto vert = vertices[vertIdx];
+      // FOR VERT-BASED OBB CALC
+      // only support bbs for known colors that map to semantic objects
+      vertMax[semanticID] = vertMax[semanticID].cwiseMax(vert);
+      vertMin[semanticID] = vertMin[semanticID].cwiseMin(vert);
+      vertCounts[semanticID] += 1;
+    }
+  }
+
+  // with mins/maxs per ID, map to objs
+  // give each ssdObj the values to build its OBB
+  debugMsgs.reserve(semanticIDToSSOBJidx.size());
+  for (int semanticID = 1; semanticID < semanticIDToSSOBJidx.size();
+       ++semanticID) {
+    // get object with given semantic ID
+    auto& ssdObj = *ssdObjs[semanticIDToSSOBJidx[semanticID]];
+    esp::vec3f center{};
+    esp::vec3f dims{};
+    const std::string debugStr = Cr::Utility::formatString(
+        "Semantic ID : {} : color : {} tag : {} present in {} verts | ",
+        semanticID, ssdObj.getColorAsString(), ssdObj.id(),
```

Comment:
Wait... so you made a `Color3ub` hex-printing utility (which `Debug` / `ESP_DEBUG()` knows natively) just to ... create a debug message, which should have been done with overhead-free `ESP_DEBUG()` in the first place? And same with reinventing vector component printing, which again `Debug` knows natively?

I'm not sure I understand the direction where the code is going.

Reply:
The list of strings is not necessarily just going to end up getting piped to debug.  Was going to conditionally write to file so console doesn't get slammed.